src/context.jl

```diff
@@ -79,19 +79,79 @@ function checkvaliddelim(delim)
         else
             Symbol(k) in names || throw(ArgumentError("invalid column name provided in `$argname` keyword argument: $k. Valid column names detected in the data are: $names. To ignore invalid columns names in `$argname`, pass `validate=false`"))
         end
     end
-    return
+    return nothing
 end
+function checkinvalidcolumns(vec::AbstractVector, argname, ncols, names)
+    # we generally expect `length(types) == ncols` but still want to support the case where
+    # an additional column is found later in the file and e.g. has its type given in `types`
+    length(vec) >= ncols || throw(ArgumentError("provided `$argname::AbstractVector` keyword argument doesn't match detected # of columns: `$(length(vec)) < $ncols`"))
+    return nothing
+end
+# if the argument isn't given as an AbstractDict or an AbstractVector then
+# we have no way to check it again the number of cols or the names
+checkinvalidcolumns(arg::Any, argname, ncols, names) = nothing
 
 @noinline nonconcretetypes(types) = throw(ArgumentError("Non-concrete types passed in `types` keyword argument, please provide concrete types for columns: $types"))
 
+# Create all the `Column`s and keep track of any non-standard eltypes for which we will
+# later need generate specialized parsing methods.
+# - `ncols` is the number of columns to create
+# - `types` is the user-given input
+function initialize_columns(ncols::Int, types, names, args...; validate)
+    columns = Vector{Column}(undef, ncols)
+    customtypes = Tuple{}
+    validate && checkinvalidcolumns(types, "types", ncols, names)
+    for i = 1:ncols
+        col = initialize_column(i, types, names, args...)
+        columns[i] = col
+        if nonstandardtype(col.type) !== Union{}
+            customtypes = tupcat(customtypes, nonstandardtype(col.type))
+        end
+    end
+    return columns, customtypes
+end
+
+# Create a `Column` with their eltype set using any user-provided types,
+# but without yet allocating a vector to hold the parsed results (see `allocate`)
+# - `i` is the column number e.g. i=1 for the 1st column.
+# - `types` is the user-given input
+function initialize_column(i, types::AbstractVector, names, stringtype, streaming::Bool, options)
+    # we generally expected `length(types) == ncols` but we still want to support the case
+    # where an additional column is found later in the file and wasn't in `types`
+    T = i <= length(types) ? types[i] : NeedsTypeDetection
+    return Column(T, options)
+end
+
+function initialize_column(i, types::AbstractDict, names, stringtype, streaming::Bool, options)
+    defaultT = streaming ? Union{stringtype, Missing} : NeedsTypeDetection
+    T = getordefault(types, names[i], i, defaultT)
+    col = Column(T, options)
+    return col
+end
+
+function initialize_column(i, types::Function, names, stringtype, streaming::Bool, options)
+    defaultT = streaming ? Union{stringtype, Missing} : NeedsTypeDetection
```

Comment:
this logic is repeated in a few of these methods, so i considered pulling it out but that actually leads to it being repeated a few places further apart in the code which seemed less nice than it being contained only in the methods of this one function